Mathlib/RingTheory/PowerSeries/Restricted.lean

```diff
@@ -0,0 +1,153 @@
+  if h : r = 0 then (simpa [h] using zero c) else
+  simp_rw [isRestricted_iff, norm_mul, norm_pow, Real.norm_eq_abs, abs_norm] at ⊢ hf
+  intro ε _
+  obtain ⟨n, hn⟩ := hf (ε / ‖r‖) (by positivity)
+  refine ⟨n, fun N hN ↦ ?_⟩
+  calc _ ≤ ‖r‖ * ‖(coeff R N) f‖ * |c| ^ N :=
+        mul_le_mul_of_nonneg (norm_mul_le _ _) (by simp) (by simp) (by simp)
+       _ < ‖r‖ * (ε / ‖r‖) := by
+        rw [mul_assoc]; aesop
+       _ = ε := mul_div_cancel₀ _ (by aesop)
+
+
+/-- The set of `‖coeff R i f‖ * c ^ i` for a given power series `f` and parameter `c`. -/
+def convergenceSet (f : PowerSeries R) : Set ℝ := {‖coeff R i f‖ * c^i | i : ℕ}
+
+open Finset in
+lemma convergenceSet_BddAbove {f : PowerSeries R} (hf : IsRestricted c f) :
+  BddAbove (convergenceSet c f) := by
+  simp_rw [isRestricted_iff] at hf
+  obtain ⟨N, hf⟩ := by simpa using (hf 1)
+  rw [bddAbove_def, convergenceSet]
+  use max 1 (max' (image (fun i ↦ ‖coeff R i f‖ * c ^ i) (range (N + 1))) (by simp))
+  simp only [Set.mem_setOf_eq, le_sup_iff, forall_exists_index, forall_apply_eq_imp_iff]
+  intro i
+  rcases le_total i N with h | h
+  · right
+    apply le_max'
+    simp only [mem_image, mem_range]
+    exact ⟨i, Order.lt_add_one_iff.mpr h, rfl⟩
+  · left
+    calc _ ≤ ‖(coeff R i) f‖ * |c ^ i| := by bound
+         _ ≤ 1 := by simpa using (hf i h).le
+
+variable [IsUltrametricDist R]
+
+open IsUltrametricDist
+
+lemma mul {f g : PowerSeries R} (hf : IsRestricted c f) (hg : IsRestricted c g) :
+    IsRestricted c (f * g) := by
+  obtain ⟨a, ha, fBound1⟩ := (bddAbove_iff_exists_ge 1).mp (convergenceSet_BddAbove _
+    ((isRestricted_iff_abs c f).mp hf))
+  obtain ⟨b, hb, gBound1⟩ := (bddAbove_iff_exists_ge 1).mp (convergenceSet_BddAbove _
+    ((isRestricted_iff_abs c g).mp hg))
+  simp only [convergenceSet, Set.mem_setOf_eq, forall_exists_index, forall_apply_eq_imp_iff]
+    at fBound1 gBound1
+  simp only [isRestricted_iff, norm_mul, norm_pow, Real.norm_eq_abs, abs_norm,
+    PowerSeries.coeff_mul] at ⊢ hf hg
+  intro ε hε
+  obtain ⟨Nf, fBound2⟩ := (hf (ε / (max a b))) (by positivity)
+  obtain ⟨Ng, gBound2⟩ := (hg (ε / (max a b))) (by positivity)
+  refine ⟨2 * max Nf Ng, fun n hn ↦ ?_⟩
+  obtain ⟨⟨fst, snd⟩, hi, ultrametric⟩ := exists_norm_finset_sum_le (Finset.antidiagonal n)
+    (fun a ↦ (coeff R a.1) f * (coeff R a.2) g)
+  obtain ⟨rfl⟩ := by simpa using hi (⟨(0, n), by simp⟩)
+  calc _ ≤ ‖(coeff R fst) f * (coeff R snd) g‖ * |c| ^ (fst + snd) := by bound
+       _ ≤ ‖(coeff R fst) f‖ * |c| ^ fst * (‖(coeff R snd) g‖ * |c| ^ snd) := by
+        grw [norm_mul_le]; grind
+  have : max Nf Ng ≤ fst ∨ max Nf Ng ≤ snd := by omega
+  rcases this with this | this
+  · calc _ < ε / max a b * b := by
+          grw [gBound1 snd]
+          gcongr
+          exact fBound2 fst (by omega)
+         _ ≤ ε := by
+          rw [div_mul_comm]
+          simpa only [div_mul_comm] using ((mul_le_iff_le_one_left hε).mpr (by bound))
```

Comment:
```suggestion
          rw [div_mul_comm, mul_le_iff_le_one_left ‹_›]
          bound
```
I hope that in the near future `field_simp` will be able to solve such goals for us...